crates/otspec/src/layout/common.rs

```diff
@@ -0,0 +1,215 @@
+        uint16 markClass
+        Offset16(Anchor) markAnchor
+    }
+);
+
+impl Debug for ScriptRecord {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if f.alternate() {
+            write!(f, "{} => {:#?}", self.scriptTag, self.scriptOffset.link)
+        } else {
+            write!(f, "{} => {:?}", self.scriptTag, self.scriptOffset.link)
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Clone)]
+/// Feature parameter data.
+///
+/// Certain OpenType features may have various ancillary data attached to them.
+/// The format of this data varies from feature to feature, so this container
+/// wraps the general concept of feature parameter data.
+pub enum FeatureParams {
+    /// The stylistic set features (`ss01`-`ss20`) may provide two parameters: a
+    /// parameter data version, currently set to zero, and a name table ID
+    /// which is used to display the stylistic set name to the user.
+    StylisticSet(uint16, uint16),
+    /// Feature parameter information for the `size` feature, including the
+    /// design size, subfamily identifier and name ID, and largest and smallest
+    /// intended sizes. This has been superseded by the `STAT` table.
+    SizeFeature(sizeFeatureParams),
+    /// The character variant features (`cv01`-`cv99`) provide various name
+    /// parameters to display information to the user.
+    CharacterVariant(cvFeatureParams),
+}
+
+bitflags! {
+    /// Lookup qualifiers
+    #[derive(Serialize, Deserialize)]
+    pub struct LookupFlags: u16 {
+        /// Position the last glyph of a cursive positioning sequence on the baseline
+        const RIGHT_TO_LEFT = 0x0001;
+        /// Skip over base glyphs
+        const IGNORE_BASE_GLYPHS = 0x0002;
+        /// Skip over ligatures
+        const IGNORE_LIGATURES = 0x0004;
+        /// Skip over all combining marks
+        const IGNORE_MARKS = 0x0008;
+        /// Indicates that the lookup table structure is followed by a MarkFilteringSet field
+        const USE_MARK_FILTERING_SET = 0x0010;
+        /// Mask off the high bits to reveal a mark class defined in the GDEF table
+        const MARK_ATTACHMENT_TYPE_MASK = 0xFF00;
+    }
+}
+
+impl Default for LookupFlags {
+    fn default() -> Self {
+        LookupFlags::empty()
+    }
+}
+
+#[allow(missing_docs, non_snake_case, non_camel_case_types)]
+#[derive(Debug, Serialize)]
+pub struct gsubgpos {
+    pub majorVersion: uint16,
+    pub minorVersion: uint16,
+    pub scriptList: Offset16<ScriptList>,
+    pub featureList: Offset16<FeatureList>,
+    pub lookupList: Offset16<LookupList>,
+}
+
+// We have to do horrible things for the Lookup table because it has
+// a heterogenous subtable vec field.
+#[allow(missing_docs, non_snake_case, non_camel_case_types)]
+#[derive(Debug)]
+pub struct Lookup {
+    pub lookupType: uint16,
+    pub flags: LookupFlags,
+    pub subtables: Vec<Box<dyn OffsetMarkerTrait>>,
+    pub mark_filtering_set: Option<uint16>,
+}
```

Comment:
Although it would mean a bunch of duplication, I'm wondering if it might be worth biting the bullet and just separating GPOS & GSUB? We could reuse a lot of the serialization/deserialization code, and binary size would be my main argument against this...

Reply:
I’m working on this at the moment. I think you’re probably right, though: separate GPOS and GSUB structs will be the way to go.